### PR TITLE
Private experimental cross-module selectors and actions

### PR DESCRIPTION
**Pull request by adamziel**
## What?

This PR introduces a private selectors APIs in `@wordpress/data` via [the new `@wordpress/experimental`](https://github.com/WordPress/gutenberg/pull/43386#issuecomment-1260379417) package:

```js
// Package wordpress/block-data:
import { unlock } from '../experiments';
import { experiments as dataExperiments } from '@wordpress/data';
const { registerPrivateActionsAndSelectors } = unlock( dataExperiments );

import { store as blockEditorStore } from './store';
import { __unstableSelectionHasUnmergeableBlock } from './store/selectors';
registerPrivateActionsAndSelectors( store, {}, {
     __experimentalHasContentRoleAttribute
} );

// plain usage
unlock( registry.select( blockEditorStore ) ).getContentLockingParent();

// usage in React
useSelect( ( select ) => ( {
  parent: privateOf( select( blockEditorStore ) ).__unstableSelectionHasUnmergeableBlock();
} ) );
```

This PR demonstrates the usage by making the `__unstableSelectionHasUnmergeableBlock` selector private. I intend to roll back this change before merging, and then follow-up with a migration PR to ensure new experimental selectors won't be added to the public API.

## Why?

Exporting `__unstable` and `__experimental` selectors [makes them available to extenders and locks them in as a part of the public API](https://github.com/WordPress/gutenberg/issues/40316). Let's not add new such exports if it can be avoided.

## Testing

Confirm the new API makes sense and that all the tests pass.

cc @jorgefilipecosta @ellatrix @talldan @draganescu @peterwilsoncc @jsnajdr

### CODE REVIEW

File: packages/block-editor/src/experiments.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+export const { register, unlock } =
+	__dangerousOptInToUnstableAPIsOnlyForCoreModules(
+		'I know using unstable features means my plugin or theme will inevitably break on the next WordPress release.',
+		'@wordpress/block-editor'
+	);
+
+const { __experimentalPrivateSelector } = unlock( dataExperiments );
+
+export const __experimentalAccessKey = register( {
+	// Follow-up on https://github.com/WordPress/gutenberg/pull/42934
```
Comment (adamziel) [inline packages/block-editor/src/experiments.js]: This comment can be removed
```suggestion
```

File: packages/block-editor/src/experiments.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	__unstableSelectionHasUnmergeableBlock: __experimentalPrivateSelector(
+		store,
+		__unstableSelectionHasUnmergeableBlock
+	),
+
+	// Follow-up on https://github.com/WordPress/gutenberg/pull/43037
```
Comment (adamziel) [inline packages/block-editor/src/experiments.js]: This comment can be removed
```suggestion
```

File: packages/data/src/index.js [hunk 3]
```diff
@@ -229,3 +234,24 @@ export const use = defaultRegistry.use;
...
+function __experimentalPrivateSelector( { name }, selector ) {
+	return ( ...args ) =>
+		selector( defaultRegistry.stores[ name ].store.getState(), ...args );
+}
+function __experimentalPrivateActionCreator( { name }, actionCreator ) {
+	return ( ...args ) =>
+		defaultRegistry.stores[ name ].store.dispatch(
+			actionCreator( ...args )
+		);
+}
```
Comment (adamziel) [inline packages/data/src/index.js]: This is beyond the scope of this PR
```suggestion
```

File: packages/data/src/index.js [hunk 4]
```diff
@@ -229,3 +234,24 @@ export const use = defaultRegistry.use;
...
 export const register = defaultRegistry.register;
+
+const experimentalAPIs = __dangerousOptInToUnstableAPIsOnlyForCoreModules(
+	'I know using unstable features means my plugin or theme will inevitably break on the next WordPress release.',
+	'@wordpress/data'
+);
+
+function __experimentalPrivateSelector( { name }, selector ) {
+	return ( ...args ) =>
+		selector( defaultRegistry.stores[ name ].store.getState(), ...args );
```
Comment (jsnajdr) [inline packages/data/src/index.js]: It is a major shortcoming that the selector always selects from the default registry, and not the one that's in React context and is returned by `useRegistry`. I wonder if we can do anything about it?

Comment (jsnajdr) [reply packages/data/src/index.js]:
```js
// registration
registerPrivateSelectors( blockEditorStore, {
  getContentLockingParent
} );

// plain usage
const { privateOf } = unlock( dataExperiments );
privateOf( registry.select( blockEditorStore ) ).getContentLockingParent();

// usage in React
useSelect( ( select ) => ( {
  parent: privateOf( select( blockEditorStore ) ).getContentLockingParent();
} ) );
```
I can register private selectors and actions on a store, and then the object returned by `select` has the good old public methods, but can also be "unlocked" with `privateOf` to access the private parts.

The same can be done with actions and `dispatch`.

Ideally, I would not need a custom `privateOf` function, but could use the general purpose `unlock` to unlock any object. Could be implemented by having a `Symbol.for( 'accessKey' )` property on an object. Very similar to how, to a `for of` loop, you can pass either an object with a `Symbol.iterator` field or an actual iterator.

**Review by jorgefilipecosta — COMMENTED** [review]
This approach seems like a solution that we could use 👍 I agree with @jsnajdr that using only the default registry is not ideal, so Ideally I think we would apply the solution @jsnajdr referred.

File: packages/data/src/index.js (hunk 4 above)
Comment (adamziel) [reply packages/data/src/index.js]: @jsnajdr I really like this idea and I'm exploring an implementation. Would you mind taking a look? I think the new lockObj/unlockObj API could very well replace the old register/unlock API – I just need some better names.

File: packages/data/src/redux-store/index.js [hunk 5]
```diff
@@ -159,6 +165,18 @@ export default function createReduxStore( key, options ) {
...
+				mapSelectors(
+					mapValues(
+						privateSelectors,
+						( selector ) =>
+							( state, ...args ) =>
+								selector( state.root, ...args )
+					),
+					store
+				);
+
```
Comment (adamziel) [inline packages/data/src/redux-store/index.js]: or perhaps

```js
configureLockingBehavior(selectors, {
    uniqueKey: experimentsPointer,
    lazyDecorator: ( privateSelectors ) => // ...
});
```

File: packages/experiments/src/index.js [hunk 6]
```diff
@@ -70,16 +88,26 @@ export const __dangerousOptInToUnstableAPIsOnlyForCoreModules = (
...
+				if ( experiment.accessKey === accessKeyOrObject ) {
+					return experiment.apisProxy;
 				}
 			}
 
 			throw new Error(
 				'There is no registered module matching the specified access key'
 			);
 		},
+		lockObj,
```
Comment (adamziel) [inline packages/experiments/src/index.js]: This could perhaps replace the register/unlock flow entirely

File: packages/experiments/src/index.js [hunk 7]
```diff
@@ -27,6 +27,12 @@ const registeredExperiments = {};
 const requiredConsent =
 	'I know using unstable features means my plugin or theme will inevitably break on the next WordPress release.';
 
+const privateAPIs = new WeakMap();
+export const experimentsSymbol = Symbol( 'experiments' );
```
Comment (adamziel) [inline packages/experiments/src/index.js]: I didn't commit all the changes to this file, this is missing:

```suggestion
export const experimentsSymbol = Symbol( 'experiments' );
export const experimentalAPIsDecorator = Symbol( 'experimentalAPIsDecorator' )
```

File: packages/data/src/index.js (hunk 4 above)
Comment (jsnajdr) [reply packages/data/src/index.js]: I'm adding this to my todo-list after finishing the React 18 migration. Which currrently utilizes me 100% 🙂

**Comment by adamziel** [comment]
I just updated this PR as @jsnajdr [suggested](https://github.com/WordPress/gutenberg/pull/44521#discussion_r983306757):

```js
// registration
registerPrivateSelectors( blockEditorStore, {
  getContentLockingParent
} );

// plain usage
unlock( registry.select( blockEditorStore ) ).getContentLockingParent();

// usage in React
useSelect( ( select ) => ( {
  parent: unlock( select( blockEditorStore ) ).getContentLockingParent();
} ) );
```

You can register private selectors and actions on a store, and the object returned by `select()` has the good old public methods, but can also be "unlocked" to access the private parts. The same can be done with actions and dispatch.

This now works with all the registries, not just the default one.

It's easy to migrate the current code to the new one – we just need to add `unlock()` where a private selector or action is used.

**Comment by adamziel** [comment]
The JavaScript unit tests seems to fail due to `Browserslist: caniuse-lite is outdated. Please run: npx browserslist@latest --update-db`, I don't see any actual failure. Any ideas how to fix it?

**Comment by adamziel** [comment]
I'm merging this PR into https://github.com/WordPress/gutenberg/pull/46131 – the two are closely related, updating them separately takes effort, and keeping the code separate makes the dependency less clear.

File: packages/data/README.md [hunk 8]
```diff
@@ -508,6 +508,26 @@ _Returns_
...
+import { unlock } from '../experiments';
+import { experiments as dataExperiments } from '@wordpress/data';
+const { registerPrivateSelectors } = unlock( dataExperiments );
+
+import { store as blockEditorStore } from './store';
+import { __unstableSelectionHasUnmergeableBlock } from './store/selectors';
+registerPrivateSelectors( store, {
+	__experimentalHasContentRoleAttribute,
+} );
+```
```
Comment (mcsf) [inline packages/data/README.md:529]: These selectors don't match, which adds confusion.

Comment (adamziel) [reply packages/data/README.md:529]: ah snap good spot, I'll fix this in #46131

Comment (adamziel) [reply packages/data/README.md:529]: https://github.com/WordPress/gutenberg/pull/46131/commits/8b5103cd63a26d5c4b8c18ec7df92f985fe3a9a4

**Pull request merged**